R/aaa-auto.R

```diff
@@ -792,7 +792,7 @@ voronoi_impl <- function(graph, generators, weights=NULL, mode=c("out", "in", "a
   res
 }
 
-get_k_shortest_paths_impl <- function(graph, weights=NULL, k, from, to, mode=c("out", "in", "all", "total")) {
+get_k_shortest_paths_impl <- function(graph, from, to, k, weights=NULL, mode=c("out", "in", "all", "total")) {
```

Comment:
The following would be safer:

```suggestion
get_k_shortest_paths_impl <- function(graph, from, to, ..., k, weights=NULL, mode=c("out", "in", "all", "total")) {
  check_dots_empty()
```

This would require users to supply names for `k` and the other optional arguments. Positional matching makes it harder:

- for users to skim the code (what does that fifth argument mean again?),
- for us to add new arguments (all new ones must go at the end to maintain full compatibility, which is not always desired).

I wonder if this could be part of Stimulus, or if we create a wrapper here (and in many other places).

Reply:
Stimulus does not have a way to indicate a preference for keyword arguments vs positional arguments, but I guess it could be added. Python does it by adding a lone `*` in the argument list (e.g., `def compare(a, b, *, key=None):` makes `key` a keyword argument only while `a` and `b` could be defined either as positional or as keyword arguments). We could adapt something similar in Stimulus if this is a concept that seems to apply to multiple languages. Languages without kwarg support could ignore the asterisk in the argument list.

Another alternative would be to add a Stimulus keyword that specifies the name of the first argument that should be treated as keyword-only (it would be `k` here), but this could cause problems if a higher level interface decides to re-order arguments.

Reply:
I like the star notation. Can you please propose something, @ntamas?

Reply:
I'll come up with a patch in Stimulus and we'll see how the idea works out.

Reply:
See the Stimulus revision tagged with 0.21.0.

There are now three alternative ways to indicate keyword-only arguments:

* You can add `KW` in front of the type of the argument in the `PARAMS` key to mark an argument as keyword-only.
* You can add `*` in the middle of the argument list at the point where the keyword-only arguments start.
* You can add a key named `FIRST_KW_PARAM` next to `PARAMS` and it must contain the name of the first keyword-only argument. This is primarily meant for backward compatibility with Stimulus 0.20 and earlier as the other two options would fail on earlier versions of Stimulus.

I tried to update the RR code generator to respect keyword-only arguments, but I haven't tested it. If it does not work, look in `RRCodeGenerator`, this is where the magic happens.

Next up would be to update `functions.yaml` in the C core to indicate which arguments should be keyword-only in generated interfaces.

Reply:
If you want to make `k` a mandatory keyword argument, I would rather specify this as a custom parameter list for the R interface, and also give a default of `k=1`. _In this specific case_, I wouldn't mark it as mandatory keyword in the C core.  Shall I go ahead?

While this feels natural in R, it wouldn't be so in Mathematica. Not sure about Python. Note that when using this function, people would always want to specify `k`.

Reply:
@ntamas How does the `*` work with parameter reordering? Is it allowed to be used in `PARAM_ORDER`?

Reply:
> Is it allowed to be used in PARAM_ORDER?

No, only in `PARAMS`. It's a neat idea, though. I'll see what can be done.

> How does the `*` work with parameter reordering?

In the parsed parameter specification, each parameter has a property named `is_keyword_only` that specifies whether the parameter is to be used as a keyword argument only if the language supports keyword arguments. The `*` is essentially a shortcut for specifying this for an argument and all subsequent arguments automatically. This comes before the parameter reordering.

Note that in theory, one could end up in a situation where keyword-only and non-keyword-only arguments are intermixed in the parameter ordering. It is up to the language-specific code generator to deal with such a situation. The RR code generator collects non-keyword-only args first and keyword-only args second (but respects `PARAM_ORDER` within each sublist). The Python code generator throws an error because it was easier to do it that way.

Reply:
`PARAM_ORDER` now supports `*` from Stimulus 0.21.2. FWIW it also supports `...`, which means "all the parameters that have not been listed explicitly so far, in their original order".